### PR TITLE
[WIP] [3.x] Changes in support of AR support on Android

### PR DESCRIPTION
**Pull request by BastiaanOlij**
This is an updated version of #41026 changed to work on the latest 3.x branch. Basically all of Fredias changes as they were with a couple split out into #47451 and a few texture lookup changes on hold (have to have a discussion with @reduz whether this is the correct approach). This thus replaces 41026 completely. Once 47451 is merged I will make the required changes in https://github.com/godotvr/godot_arcore and see if we can finally make this mergeable.

### CODE REVIEW

**Review by m4gr3d — DISMISSED** [review]

**Comment by BastiaanOlij** [comment]
@m4gr3d I'm putting a little bit more time into ARCore, seeing if I can finally get it to run. I just rebased this PR and cleaned up all the merge conflicts. I'm not sure if we lost something along the way or if this is because ARCore is now a plugin, but is it still necessary to have ARCore as an `xr_mode` ? I think all important bits were moved to the AAR but I could be wrong (note that currently the plugin won't load and it may be because we lost something here).

**Comment by m4gr3d** [comment]
> @m4gr3d I'm putting a little bit more time into ARCore, seeing if I can finally get it to run. I just rebased this PR and cleaned up all the merge conflicts. I'm not sure if we lost something along the way or if this is because ARCore is now a plugin, but is it still necessary to have ARCore as an `xr_mode` ? I think all important bits were moved to the AAR but I could be wrong (note that currently the plugin won't load and it may be because we lost something here).

@BastiaanOlij Yes it's still necessary to have ARCore as an `xr_mode`. They are a couple of edits to the manifest that are needed when using ARCore, so we use the `xr_mode` to figure out when to add them.

Let me take another look at the PR seeing if I can spot anything else that's awry.

**Comment by BastiaanOlij** [comment]
> @BastiaanOlij Yes it's still necessary to have ARCore as an `xr_mode`. They are a couple of edits to the manifest that are needed when using ARCore, so we use the `xr_mode` to figure out when to add them.
> 
> Let me take another look at the PR seeing if I can spot anything else that's awry.

We might have lost those along the way, the manifest changes we need are in the AAR manifest though.
That said, I do wonder if that has something to do with the plugin not loading?

File: platform/android/java/lib/src/org/godotengine/godot/xr/XRMode.java [hunk 1]
```diff
@@ -36,7 +36,8 @@
 public enum XRMode {
 	REGULAR(0, "Regular", "--xr_mode_regular", "Default Android Gamepad"), // Regular/flatscreen
 	OVR(1, "Oculus Mobile VrApi (Deprecated)", "--xr_mode_ovr", ""),
-	OPENXR(2, "OpenXR", "--xr_mode_openxr", "");
+	OPENXR(2, "OpenXR", "--xr_mode_openxr", ""),
+	ARCORE(3, "AR Core", "--xr_mode_arcore", "");
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/xr/XRMode.java:40]: The third argument should be `Default Android Gamepad`, same as the `REGULAR` option.

File: platform/android/java/lib/src/org/godotengine/godot/Godot.java [hunk 2]
```diff
@@ -747,6 +754,10 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle ic
 		return containerLayout;
 	}
 
+	public XRMode getXRMode() {
+		return this.xrMode;
+	}
```
Comment (m4gr3d) [inline platform/android/java/lib/src/org/godotengine/godot/Godot.java:759]: What's the use for this getter?

Comment (BastiaanOlij) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java:759]: I think this is called in the plugins? this is code from you that I've ported along for the past 3 years so might just be obsolete :)

Comment (BastiaanOlij) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java]: Removed it

File: platform/android/java/lib/src/org/godotengine/godot/xr/XRMode.java (hunk 1 above)
Comment (BastiaanOlij) [reply platform/android/java/lib/src/org/godotengine/godot/xr/XRMode.java]: Changed it :)

File: platform/android/java/lib/src/org/godotengine/godot/Godot.java (hunk 2 above)
Comment (BastiaanOlij) [reply platform/android/java/lib/src/org/godotengine/godot/Godot.java]: Turns out this was needed afterall: https://github.com/GodotVR/godot_arcore/blob/make_this_work/plugin/src/main/java/org/godotengine/plugin/arcore/ARCorePlugin.java#L20

**Comment by SmartySmart702** [comment]
What is the current state of arcore in Godot? Is someone still working on this?

**Comment by BastiaanOlij** [comment]
> What is the current state of arcore in Godot? Is someone still working on this?

Sadly this has been at a standstill but I'm trying to get some interest in seeing if we can finally push it over the finish line.

File: modules/gdnative/gdnative_api.json [hunk 3]
```diff
@@ -6580,7 +6580,23 @@
         "major": 1,
         "minor": 1
       },
-      "next": null,
+      "next": {
```
Comment (salvprest) [inline modules/gdnative/gdnative_api.json:6583]: Im trying to make this work too, in this line a sub android api major:1 minor:2 is introduced, and the godot_android_get _display_rotation is defined under this sub api,
in the original work PR https://github.com/godotengine/godot/pull/26221 instead the original android api major:1 minor:1 is modified and the godot_android_get_display_rotation is defined at the same level of the other android api call,
so I think that is an error

**Comment by SmartySmart702** [comment]
are there plans to get this ready for 4.1?

**Comment by BastiaanOlij** [comment]
@SmartySmart702 we should first get it to work on Godot 3... 

That said, anyone interested in implementing ARCore for Godot 4 would be very welcome to do so

**Comment by dsnopek** [comment]
Should we just close this at this point? This could still theoretically get fixed up, but I don't know if @BastiaanOlij is still interested in this, and if not, someone else could open a new PR

**Comment by BastiaanOlij** [comment]
@dsnopek yeah I'll close it, it's been implemented now with the work recently merged.

**Pull request closed without merging**